src/java/org/apache/cassandra/index/sai/disk/PostingList.java

```diff
@@ -60,4 +61,77 @@ default void close() throws IOException {}
      * @return first segment row ID which is >= the target row ID or {@link PostingList#END_OF_STREAM} if one does not exist
      */
     int advance(int targetRowID) throws IOException;
+
+    /**
+     * @return peekable wrapper of current posting list
+     */
+    default PeekablePostingList peekable()
+    {
+        return new PeekablePostingList(this);
+    }
+
+    public static class PeekablePostingList implements PostingList
+    {
+        private final PostingList wrapped;
+
+        private boolean peeked = false;
+        private int next;
+
+        public PeekablePostingList(PostingList wrapped)
+        {
+            this.wrapped = wrapped;
+        }
+
+        public int peek()
+        {
+            if (peeked)
+                return next;
+
+            try
+            {
+                peeked = true;
+                return next = wrapped.nextPosting();
+            }
+            catch (IOException e)
+            {
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        @Override
+        public int nextPosting() throws IOException
+        {
+            if (peeked)
+            {
+                peeked = false;
+                return next;
+            }
+            return wrapped.nextPosting();
+        }
+
+        @Override
+        public int size()
+        {
+            return wrapped.size();
+        }
+
+        @Override
+        public int advance(int targetRowID) throws IOException
```

Comment:
General question about the navigating through PostingLists. Is there ever a case where we would want to "reset" our position in a PostingList? 

For example, if we advance to a targetRowID that extends beyond the current range, we'll return an END_OF_STREAM value, and any further calls to peek or nextPosting would return the same. Is there anyway to back out of that?

I guess one way would be to implement something like `PeekingPostingList#advanceWithoutConsuming`, which relates to my question above.

Reply:
> General question about the navigating through PostingLists. Is there ever a case where we would want to "reset" our position in a PostingList?

`RangeUnionIterator` and `RangeIntersectionIterator` don't iterate backward as we return data in partition order (eg. within the scope of one sstable, targetRowID should be monotonic growing)

Reply:
Added `advanceWithoutConsuming` to `PeekablePostingList` for `MergePostingList`